diff -ruN android_kernel_xiaomi_sm6150/drivers/Kconfig android_kernel_xiaomi_sm6150-test/drivers/Kconfig
--- android_kernel_xiaomi_sm6150/drivers/Kconfig	2025-12-26 21:55:05.292814397 +0800
+++ android_kernel_xiaomi_sm6150-test/drivers/Kconfig	2025-12-27 10:00:00.112135683 +0800
@@ -229,4 +229,6 @@
 
 source "drivers/kprofiles/Kconfig"
 
+source "drivers/kernelsu/Kconfig"
+
 endmenu
diff -ruN android_kernel_xiaomi_sm6150/drivers/Makefile android_kernel_xiaomi_sm6150-test/drivers/Makefile
--- android_kernel_xiaomi_sm6150/drivers/Makefile	2025-12-26 21:55:05.294607701 +0800
+++ android_kernel_xiaomi_sm6150-test/drivers/Makefile	2025-12-27 10:00:18.864430898 +0800
@@ -193,3 +193,5 @@
 obj-$(CONFIG_GNSS)		+= gnss/
+
+obj-$(CONFIG_KSU) += kernelsu/
