diff -ruN android_kernel_xiaomi_sm6150/fs/internal.h android_kernel_xiaomi_sm6150-test/fs/internal.h
--- android_kernel_xiaomi_sm6150/fs/internal.h	2025-12-26 21:55:07.461251334 +0800
+++ android_kernel_xiaomi_sm6150-test/fs/internal.h	2025-12-27 01:26:55.771532375 +0800
@@ -69,6 +69,8 @@
 
 extern void __init mnt_init(void);
 
+int path_umount(struct path *path, int flags);
+
 extern int __mnt_want_write(struct vfsmount *);
 extern int __mnt_want_write_file(struct file *);
 extern int mnt_want_write_file_path(struct file *);
diff -ruN android_kernel_xiaomi_sm6150/fs/namespace.c android_kernel_xiaomi_sm6150-test/fs/namespace.c
--- android_kernel_xiaomi_sm6150/fs/namespace.c	2025-12-26 21:55:07.471132923 +0800
+++ android_kernel_xiaomi_sm6150-test/fs/namespace.c	2025-12-27 01:23:18.304467506 +0800
@@ -1832,6 +1832,36 @@
 	       dentry->d_fsdata == &mntns_operations;
 }
 
+static int can_umount(const struct path *path, int flags)
+{
+    struct mount *mnt = real_mount(path->mnt);
+    if (flags & ~(MNT_FORCE | MNT_DETACH | MNT_EXPIRE | UMOUNT_NOFOLLOW))
+        return -EINVAL;
+    if (!may_mount())
+        return -EPERM;
+    if (path->dentry != path->mnt->mnt_root)
+        return -EINVAL;
+    if (!check_mnt(mnt))
+        return -EINVAL;
+    if (mnt->mnt.mnt_flags & MNT_LOCKED)
+        return -EINVAL;
+    if (flags & MNT_FORCE && !capable(CAP_SYS_ADMIN))
+        return -EPERM;
+    return 0;
+}
+
+int path_umount(struct path *path, int flags)
+{
+    struct mount *mnt = real_mount(path->mnt);
+    int ret;
+	ret = can_umount(path, flags);
+    if (!ret)
+        ret = do_umount(mnt, flags);
+    dput(path->dentry);
+	mntput_no_expire(mnt);
+	return ret;
+}
+
 struct mnt_namespace *to_mnt_ns(struct ns_common *ns)
 {
 	return container_of(ns, struct mnt_namespace, ns);
